org.jacoco.core/src/org/jacoco/core/internal/instr/Companions.java

```diff
@@ -0,0 +1,267 @@
+/*******************************************************************************
+ * Copyright (c) 2009, 2016 Mountainminds GmbH & Co. KG and Contributors
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *    Evgeny Mandrikov - initial API and implementation
+ *
+ *******************************************************************************/
+package org.jacoco.core.internal.instr;
+
+import java.lang.ref.WeakReference;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.Map;
+import java.util.WeakHashMap;
+
+import org.jacoco.core.internal.Java9Support;
+import org.jacoco.core.internal.data.CRC64;
+import org.jacoco.core.internal.flow.ClassProbesAdapter;
+import org.jacoco.core.runtime.RuntimeData;
+import org.objectweb.asm.ClassReader;
+import org.objectweb.asm.ClassVisitor;
+import org.objectweb.asm.ClassWriter;
+import org.objectweb.asm.Label;
+import org.objectweb.asm.MethodVisitor;
+import org.objectweb.asm.Opcodes;
+import org.objectweb.asm.Type;
+
+/**
+ * Instrumentation with creation of "companion" classes in "online" mode, which
+ * will hold state of instrumented classes. This allows to avoid addition of
+ * members to classes.
+ */
+public class Companions {
+
+	/**
+	 * {@link ClassLoader#defineClass(String, byte[], int, int)}
+	 */
+	private static final Method DEFINE_CLASS;
+
+	static {
+		try {
+			DEFINE_CLASS = ClassLoader.class.getDeclaredMethod("defineClass",
+					String.class, byte[].class, Integer.TYPE, Integer.TYPE);
+			DEFINE_CLASS.setAccessible(true);
```

Comment:
@bjkail Well most likely you right and [#ReflectiveAccessByInstrumentationAgents](http://openjdk.java.net/projects/jigsaw/spec/issues/#ReflectiveAccessByInstrumentationAgents) ( [`Instrumentation.redefineModule`](http://download.java.net/java/jdk9/docs/api/java/lang/instrument/Instrumentation.html#redefineModule-java.lang.reflect.Module-java.util.Set-java.util.Map-java.util.Map-java.util.Set-java.util.Map-) ) is a way to go. To be honest - I also didn't followed story behind this closely enough to give more details, because we don't use reflection in master. In any case fact that things change on this subject in JDK 9, increases risks for incorporation of such use of reflection.

> I seem to recall javaagent do not have their own Module, so perhaps one would have to be dynamically created that contains code to call setAccessible, and then the Module would be specified in extraOpens for "java.lang".

I assume that simply [`Class.getModule`](http://download.java.net/java/jdk9/docs/api/java/lang/Class.html#getModule--) will do the job.

I don't know what to ask from Rory - I would prefer to dig deeper into a subject and/or try to use to get some concrete questions, but have doubts about my free time in short-term. Also suspect that as an answer you'll be redirected to one of OpenJDK mailing lists :wink: However don't hesitate to ask.